GCEWindowsAgent/serial_listener.go

```diff
@@ -0,0 +1,462 @@
+		"snapshot-pre-snapshot-script-url",
+		"snapshot-post-snapshot-script-url"}
+
+	storageURL = "storage.googleapis.com"
+
+	bucket = `([a-z0-9][-_.a-z0-9]*)`
+	object = `(.+)`
+	// Many of the Google Storage URLs are supported below.
+	// It is preferred that customers specify their object using
+	// its gs://<bucket>/<object> URL.
+	bucketRegex = regexp.MustCompile(fmt.Sprintf(`^gs://%s/?$`, bucket))
+	gsRegex     = regexp.MustCompile(fmt.Sprintf(`^gs://%s/%s$`, bucket, object))
+	// Check for the Google Storage URLs:
+	// http://<bucket>.storage.googleapis.com/<object>
+	// https://<bucket>.storage.googleapis.com/<object>
+	gsHTTPRegex1 = regexp.MustCompile(fmt.Sprintf(`^http[s]?://%s\.storage\.googleapis\.com/%s$`, bucket, object))
+	// http://storage.cloud.google.com/<bucket>/<object>
+	// https://storage.cloud.google.com/<bucket>/<object>
+	gsHTTPRegex2 = regexp.MustCompile(fmt.Sprintf(`^http[s]?://storage\.cloud\.google\.com/%s/%s$`, bucket, object))
+	// Check for the other possible Google Storage URLs:
+	// http://storage.googleapis.com/<bucket>/<object>
+	// https://storage.googleapis.com/<bucket>/<object>
+	//
+	// The following are deprecated but checked:
+	// http://commondatastorage.googleapis.com/<bucket>/<object>
+	// https://commondatastorage.googleapis.com/<bucket>/<object>
+	gsHTTPRegex3 = regexp.MustCompile(fmt.Sprintf(`^http[s]?://(?:commondata)?storage\.googleapis\.com/%s/%s$`, bucket, object))
+)
+
+func clearCurrentMsg() {
+	currentMsg = nil
+}
+
+func parseMessage() *serialRequest {
+	defer clearCurrentMsg()
+	msg := serialRequest{}
+	err := json.Unmarshal(currentMsg, &msg)
+	if err != nil {
+		logger.Fatalf(err.Error())
+	}
+	return &msg
+}
+
+func handlePrepareForSnapshotRequest(req serialRequest) {
+	if seenPreSnapshotoperationIds.Contains(req.OperationId) {
+		logger.Infof("ignoring pre-snapshot request with operation id %d", req.OperationId)
+		return
+	}
+
+	seenPreSnapshotoperationIds.Add(req.OperationId, nil)
+
+	logger.Infof("running pre-snapshot script")
+	config := getConfig()
+	code := runScript(config.PreSnapshotScriptUrl, config.Timeout)
+	res := serialResponse{identifier, "PRERESP", req.Version, code, req.OperationId}
+	out, err := json.Marshal(res)
+	if err != nil {
+		logger.Fatalf(err.Error())
+	}
+	writeSerial(serialPort, out)
+}
+
+func handleResumePostSnapshotRequest(req serialRequest) {
+	if seenPostSnapshotoperationIds.Contains(req.OperationId) {
+		logger.Infof("ignoring post-snapshot request with operation id %d", req.OperationId)
+		return
+	}
+	seenPostSnapshotoperationIds.Add(req.OperationId, nil)
+
+	logger.Infof("running post-snapshot script")
+	config := getConfig()
+	code := runScript(config.PostSnapshotScriptUrl, config.Timeout)
+	res := serialResponse{identifier, "POSTRESP", req.Version, code, req.OperationId}
+	out, err := json.Marshal(res)
+	if err != nil {
+		logger.Fatalf(err.Error())
+	}
+	writeSerial(serialPort, out)
+}
+
+func handleMessage(msg serialRequest) {
```

Comment:
this func is just one switch statement and used in one place, just put the switch statement there